cpp/tensorrt_llm/thop/moeUtilOp.cpp

```diff
@@ -0,0 +1,449 @@
+            nullptr, static_cast<float const*>(unpermuted_final_scales.const_data_ptr()),
+            static_cast<int const*>(expanded_source_row_to_expanded_dest_row.const_data_ptr()),
+            static_cast<int const*>(expert_for_source_row.const_data_ptr()), num_valid_tokens_ptr, num_rows,
+            hidden_size, experts_per_token, parallelism_config, stream,
+            static_cast<__nv_bfloat16*>(final_output.data_ptr()));
+        break;
+    case torch::kHalf:
+        runMoEFinalizeScaleOp<half, half, half>(static_cast<half const*>(gemm2_output.const_data_ptr()),
+            // static_cast<half const*>(fc2_expert_biases.const_data_ptr()),
+            nullptr, static_cast<float const*>(unpermuted_final_scales.const_data_ptr()),
+            static_cast<int const*>(expanded_source_row_to_expanded_dest_row.const_data_ptr()),
+            static_cast<int const*>(expert_for_source_row.const_data_ptr()), num_valid_tokens_ptr, num_rows,
+            hidden_size, experts_per_token, parallelism_config, stream, static_cast<half*>(final_output.data_ptr()));
+        break;
+    default:
+        throw std::invalid_argument(
+            "Invalid dtype, only supports input tensor with float32, float16 and bfloat16 dtype");
+        break;
+    }
+    return final_output;
+}
+
+} // namespace torch_ext
+
+TORCH_LIBRARY_FRAGMENT(trtllm, m)
+{
+    m.def(
+        "moe_permute_op(Tensor input, Tensor token_selected_experts, Tensor? token_final_scales, Tensor "
+        "fc1_expert_weights, Tensor fc2_expert_weights, Tensor[]? quant_scales, Tensor? input_sf, int "
+        "num_experts_on_rank, int tp_size, int tp_rank, int ep_size, int ep_rank, int cluster_size, int cluster_rank, "
+        "bool min_latency_mode, bool use_fp8_block_scaling)"
+        "-> (Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor)");
+    m.def(
+        "moe_finalize_scale_op(Tensor gemm2_output, Tensor fc2_expert_biases, Tensor unpermuted_final_scales, Tensor "
+        "expanded_source_row_to_expanded_dest_row, Tensor expert_for_source_row, Tensor "
+        "expert_first_token_offset_tensor, SymInt num_rows, SymInt hidden_size, int experts_per_token, int "
+        "num_experts_per_node, int tp_size, int tp_rank, int ep_size, int ep_rank)"
+        "-> (Tensor)");
+    m.def(
+        "moe_expand_op(Tensor input, Tensor? token_final_scales, Tensor permuted_source_token_ids, int num_rows, "
+        "Tensor expert_first_token_offset_tensor, int hidden_size, int experts_per_token, int num_experts_per_node, "
+        "int tp_size, int tp_rank, int ep_size, int ep_rank, bool use_fp8_block_scaling)"
+        "-> (Tensor, Tensor, Tensor)");
+}
+
+TORCH_LIBRARY_IMPL(trtllm, CUDA, m)
+{
+    m.impl("moe_permute_op", &torch_ext::moe_permute_op);
+    m.impl("moe_finalize_scale_op", &torch_ext::run_moe_finalize_scale_op);
+    m.impl("moe_expand_op", &torch_ext::run_moe_expand_op);
```

Comment:
Is this op used?

Reply:
moe_expand_op is not used now. It is a independent op whose logic has already been included in moe_permute_op. I use it for debug before. I want to remove it when all refactoring done especially when it really not needed.